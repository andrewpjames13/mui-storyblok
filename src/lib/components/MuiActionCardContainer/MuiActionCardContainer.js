import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import {
  Grow,
  Paper,
  Popper,
  MenuList,
  Button,
  ClickAwayListener,
} from '@material-ui/core';
import { withRouter } from 'react-router-dom';
import StoryBlok from '../../utils/Storyblok';
import MuiActionCard from './components/MuiActionCard/MuiActionCard';
import MuiTypography from '../MuiTypography/MuiTypography';

export const MuiActionCardContainer = ({
  rootClass,
  menuName,
  actionCards,
  height,
  width,
}) => {
  const [open, setOpen] = useState(false);
  const anchorRef = useRef(null);
  const defaultStyling = {
    padding: '20px 10%',
    marginTop: '10px',
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'space-between',
    outline: 'none',
  };
  const styles = StoryBlok.arrayToMuiStyles(rootClass, { ...defaultStyling });

  const handleToggle = () => {
    setOpen(prevOpen => !prevOpen);
  };

  const handleClose = (event) => {
    if (anchorRef.current && anchorRef.current.contains(event.target)) {
      return;
    }

    setOpen(false);
  };

  function handleListKeyDown(event) {
    if (event.key === 'Tab') {
      event.preventDefault();
      setOpen(false);
    }
  }

  // return focus to the button when we transitioned from !open -> open
  const prevOpen = useRef(open);
  useEffect(() => {
    if (prevOpen.current === true && open === false) {
      anchorRef.current.focus();
    }

    prevOpen.current = open;
  }, [open]);

  return (
    <>
      <Button
        ref={anchorRef}
        aria-controls={open ? 'menu-list-grow' : undefined}
        aria-haspopup="true"
        onClick={handleToggle}
        data-testid="menuButton"
      >
        <MuiTypography {...menuName[0]} />
      </Button>
      <Popper
        open={open}
        anchorEl={anchorRef.current}
        role={undefined}
        transition
        disablePortal
      >
        {({ TransitionProps, placement }) => (
          <Grow
            {...TransitionProps}
            style={{
              transformOrigin:
                placement === 'bottom' ? 'center top' : 'center bottom',
            }}
          >
            <Paper style={{ minWidth: '100vw', height: '100%' }}>
              <ClickAwayListener onClickAway={handleClose}>
                <MenuList
                  autoFocusItem={open}
                  id="menu-list-grow"
                  onKeyDown={handleListKeyDown}
                  className={styles.root}
                >
                  {actionCards.map((card, index) => (
                    <MuiActionCard {...card} key={index} height={height} width={width} />
                  ))}
                </MenuList>
              </ClickAwayListener>
            </Paper>
          </Grow>
        )}
      </Popper>
    </>
  );
};

export default withRouter(MuiActionCardContainer);

MuiActionCardContainer.propTypes = {
  /**
   * storyblok multiselect of css classes
   * Override or extend the styles applied to the component
  * */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /**
   * Height of the each Action Card.
   */
  height: PropTypes.string,
  /**
   * Width of the each Action Card.
   */
  width: PropTypes.string,
  /**
   * Text Displayed for menu
   * Component: MuiTypography
   */
  menuName: PropTypes.arrayOf(PropTypes.shape({
    component: PropTypes.string.isRequired,
  })).isRequired,
  /**
   * Cards passed ot MuiActionCardContainer to render
   * Component: MuiActionCard
   */
  actionCards: PropTypes.arrayOf(PropTypes.shape({
    component: PropTypes.string.isRequired,
  })).isRequired,
};

MuiActionCardContainer.defaultProps = {
  rootClass: [],
  height: '100px',
  width: '200px',
};
