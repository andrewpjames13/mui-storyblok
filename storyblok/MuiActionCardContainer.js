module.exports = {
  "name": "MuiActionCardContainer",
  "display_name": null,
  "updated_at": "2020-06-09T16:20:01.920Z",
  "schema": {
    "rootClass": {
      "type": "options",
      "description": "* storyblok multiselect of css classes\n* Override or extend the styles applied to the component"
    },
    "menuName": {
      "type": "bloks",
      "description": "* Text Displayed for menu",
      "default_value": "",
      "restrict_components": true,
      "component_whitelist": [
        "MuiTypography"
      ],
      "required": true,
      "maximum": "1"
    },
    "actionCards": {
      "type": "bloks",
      "required": true,
      "maximum": "",
      "restrict_components": true,
      "component_whitelist": [
        "MuiActionCard"
      ]
    },
    "height": {
      "type": "text",
      "description": "* Height of the each Action Card.",
      "default_value": "100px"
    },
    "width": {
      "type": "text",
      "description": "* Width of the each Action Card.",
      "default_value": "200px"
    }
  },
  "image": null,
  "preview_field": null,
  "is_root": false,
  "preview_tmpl": null,
  "is_nestable": true,
  "all_presets": [],
  "preset_id": null,
  "real_name": "MuiActionCardContainer",
  "component_group_uuid": null
}