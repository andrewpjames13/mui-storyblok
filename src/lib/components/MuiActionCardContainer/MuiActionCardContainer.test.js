import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import { MuiActionCardContainer } from './MuiActionCardContainer';

function setup() {
  const props = {
    menuName: [{
      component: 'MuiTypography',
      content: [{
        component: 'MuiText',
        text: 'Action Card Menu',
      }],
    }],
    height: '100px',
    width: '100px',
    actionCards: [{
      component: 'MuiActionCard',
      header: [{
        component: 'MuiTypography',
        content: [{
          component: 'MuiText',
          text: 'Card Header',
        }],
      }],
      icon: [{
        component: 'MuiIcon',
        iconName: 'star',
      }],
      text: [{
        component: 'MuiTypography',
        content: [{
          component: 'MuiText',
          text: 'Action card body test',
        }],
      }],
      redirectRoute: '/page-welcome',
      history: {
        push: jest.fn(),
      },
    }],
  };
  const comp = mount(<MuiActionCardContainer {...props} />);
  return { comp, props };
}

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

global.document.createRange = () => ({
  setStart: () => {},
  setEnd: () => {},
  commonAncestorContainer: {
    nodeName: 'BODY',
    ownerDocument: document,
  },
});

describe('<MuiActionCardContainer />', () => {
  it('renders MuiActionCardContainer', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(
      <MemoryRouter>
        <MuiActionCardContainer {...props} />
      </MemoryRouter>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should handleToggle and set open to close and close to open', () => {
    const { props } = setup();
    act(() => {
      ReactDOM.render(<MuiActionCardContainer {...props} />, container);
    });

    const btn = container.querySelector('button');
    act(() => {
      btn.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(container.innerHTML.includes('Action card body test')).toEqual(true);
  });
});
