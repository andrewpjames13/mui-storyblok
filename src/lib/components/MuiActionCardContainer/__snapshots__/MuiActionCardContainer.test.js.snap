// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiActionCardContainer /> snapshot 1`] = `
<div>
  <button
    aria-haspopup="true"
    className="MuiButtonBase-root MuiButton-root MuiButton-text"
    data-testid="menuButton"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      <div
        className="MuiTypography-root makeStyles-root-8 MuiTypography-body1"
      >
        <span
          className="MuiTypography-root makeStyles-root-9 MuiTypography-body1"
        >
           Action Card Menu 
        </span>
      </div>
    </span>
  </button>
</div>
`;
